From 7a9aad46ee3c32cd4fe44a8598ebf75545b972da Mon Sep 17 00:00:00 2001
From: Dien Pham <dien.pham.ry@renesas.com>
Date: Fri, 30 Mar 2018 13:15:10 +0700
Subject: [PATCH 169/909] thermal: rcar_gen3_thermal: Use DIV_ROUND_CLOSEST
 correctly as its description

commit 5627c42a1bd5812837c8d4dfce5ac53867911556 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

As description for DIV_ROUND_CLOSEST in file include/linux/kernel.h.
  "Result is undefined for negative divisors
   if the dividend variable type is unsigned
   and for negative dividends
   if the divisor variable type is unsigned."

In current code, the FIXPT_DIV uses DIV_ROUND_CLOSEST
but has not checked sign of divisor before using.
It makes undefined temperature value in case
the value is negative.

This patch fixes to satisfy DIV_ROUND_CLOSEST description
and fix bug too.

Signed-off-by: Van Do <van.do.xw@rvc.renesas.com>
Signed-off-by: Dien Pham <dien.pham.ry@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/thermal/rcar_gen3_thermal.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/thermal/rcar_gen3_thermal.c b/drivers/thermal/rcar_gen3_thermal.c
index b27bcdec427e..cc3f605e9eeb 100644
--- a/drivers/thermal/rcar_gen3_thermal.c
+++ b/drivers/thermal/rcar_gen3_thermal.c
@@ -216,10 +216,12 @@ static int rcar_gen3_thermal_round(int temp)
 static int rcar_gen3_thermal_convert_temp(struct rcar_gen3_thermal_tsc *tsc)
 {
 	int mcelsius = 0, val1, val2;
-	u32 reg;
+	long reg;
 
 	if (is_ths_typeA) {
 		/* Read register and convert to mili Celsius */
+
+		/* Read register and convert value to signed variable type */
 		reg = rcar_gen3_thermal_read(tsc, REG_GEN3_TEMP) & CTEMP_MASK;
 
 		val1 = FIXPT_DIV(FIXPT_INT(reg) - tsc->coef.b1, tsc->coef.a1);
-- 
2.17.1

